docs/design/build-system.md

```diff
@@ -0,0 +1,154 @@
+
+The build system allows specifying replacements for specific files by replacing what path is loaded from the virtual file system.
+This is used for conditional loading of code at build time.
+
+### Stylesheets
+
+Two types of stylesheets are used in the build system: global stylesheets and component stylesheets.
+Global stylesheets are injected into the `index.html` file, while component stylesheets are loaded directly into compiled Angular components.
+
+The build system supports plain CSS stylesheets as well as the Sass, LESS and Stylus CSS pre-processors.
+Stylesheet processing functionality is provided by `sass-loader`, `less-loader`, `stylus-loader`, `postcss-loader`, `postcss-import`, augmented in the build system by custom webpack plugins.
+
+todo: style resources
+
+
+### Assets
+
+Assets in the build system refer specifically to a list of files or directories that are meant to be copied verbatim as build artifacts.
+These files are not processed and commonly include images, favicons, pdfs and other generic file types.
+They are loaded into the compilation using `copy-webpack-plugin`.
+
+
+### Scripts
+
+Scripts in the build system refer specifically to Javascript files that are meant to be loaded directly on `index.html` without being processed.
+They are loaded into the compilation using a custom webpack plugin.
+
+
+### Third party dependencies
+
+Third party dependencies are mostly inside `node_modules` and are referenced via imports in source files.
+Stylesheet third party dependencies are treated mostly the same as sources.
+
+JavaScript third party dependencies suffer a more involved process.
+They are first resolved to a folder in `node_modules` via [Node Module Resolution](https://nodejs.org/api/modules.html#modules_modules).
+Afterwards one of the entry points in that folder's `package.json` is used following the `es2015 > browser > module > main` priority list.
+
+Once the actual JavaScript file is determined, it is loaded into the compilation together with it's source map.
+
+
+## Code splitting
+
+Code is automatically split into different files (or chunks, for js files) based on a few different triggers.
+
+The main TypeScript entry point and it's dependencies are bundled into the `main` chunk.
+Global styles and scripts get one file per entry, named after themselves.
+
+JavaScript code imported only via dynamic imports is automatically split into a separate chunk that is loaded asynchronously named after the file containing the dynamic import.
+If multiple asynchronous chunks contain a reference to the same module, it is placed in a new asynchronously loaded chunk named after the other chunks that use it.
+
+There is also a special chunk called `runtime` that contains the module loading logic and is loaded before the others.
+
+
+## Optimizations
+
+The build system contains optimizations aimed and improving the performance, for development builds, or the size of artifacts, for production builds.
```

Comment:
Nit: The commas here are confusingly worded. I think you meant something like:

"The build system contains optimizations aimed **at** improving the performance **(for development builds)**, or the size of artifacts **(for production builds)**. These are often mutually exclusive and thus we cannot just default to always using them."

Reply:
Fixed